fairscale/experimental/nn/offload.py

```diff
@@ -0,0 +1,447 @@
+            and current_shard_params + sum(p.numel() for p in m.parameters()) > number_parameters_per_shard
+            and current_shard < number_splits - 1
+        ):
+            current_shard += 1
+
+        splits[current_shard].append(m)
+
+    for i, split in enumerate(splits):
+        current_shard_params = sum(p.numel() for sm in split for p in sm.parameters())
+        logging.info(f"Shard {i} holds {current_shard_params/1e6:.2f}M parameters")
+
+    return splits
+
+
+class ModelShard(nn.Module):
+    """
+    Wrap one shard of the model, make it possible to load parameters on the 
+    fly for the FW and BW pass on the given device.
+    """
+
+    def __init__(
+        self, cpu_model_shard: nn.Module, device: torch.device, offload_device: torch.device, index: int,
+    ):
+        super().__init__()
+        self.model_shard = cpu_model_shard
+        self.index = index
+
+        # Save all the parameter sizes to be able to restore them
+        self.device = device
+        torch.cuda.device(self.device)
+
+        self.offload_device = offload_device
+
+        self.model_shard.to(offload_device)
+        self.cuda_stream = torch.cuda.Stream(
+            device=self.device
+        )  # needed to make sure load/offload really run in parallel with compute
+
+    def forward(self, *inputs):  # type: ignore
+        return self.model_shard(*inputs) if isinstance(inputs, tuple) else self.model_shard(inputs)
+
+    def to(self, device: torch.device) -> "ModelShard":  # type: ignore
+        # Make sure that the lookahead and lookback shards are not captured by this call
+        self.model_shard.to(device)
+        return self
+
+    def train(self, mode: bool = True) -> "ModelShard":
+        # Make sure that the lookahead and lookback shards are not captured by this call
+        self.model_shard.train(mode)
+        return self
+
+    def to_device(self) -> None:
+        self.model_shard.to(device=self.device, non_blocking=True)
+
+    def forward_load(self, non_blocking: bool = True) -> None:
+        with torch.cuda.stream(self.cuda_stream):
+            # Restore all the parameter buffers
+            self.model_shard.to(device=self.device, non_blocking=non_blocking)
+
+    def backward_load(self, non_blocking: bool = True) -> None:
+        with torch.cuda.stream(self.cuda_stream):
+            self.model_shard.to(self.device, non_blocking=non_blocking)
+
+    def forward_drop(self, non_blocking: bool = True) -> None:
+        with torch.cuda.stream(self.cuda_stream):
+            self.model_shard.to(self.offload_device, non_blocking=non_blocking)
+
+    def backward_drop(self, non_blocking: bool = True) -> None:
+        with torch.cuda.stream(self.cuda_stream):
+            self.model_shard.to(self.offload_device, non_blocking=non_blocking)
+
+
+class ActivationCheckpointing(torch.autograd.Function):
+    """
+     This Function enables us to override the forward and backward pass of the nn.Module.
```

Comment:
the comment here doesn't seem to describe the class? In particular, it doesn't mention activation at all but the class name is activation.